### ISSUE TITLE
OutOfMemory (OOM) Crash

### ISSUE THREAD

**Issue by patrickgold** [florisboard/florisboard#677]
This issue is a collection for all OOM (OutOfMemory) related crashes, as the number of duplicates regarding this topic is getting out-of-hand. Further crash reports with a stacktracte similar to the one attached below will be marked as duplicate and linked to this issue.

In general, these are the known areas in v0.3.10 which are most likely the sources of OOM crashes and which need fixing for v0.3.11:
- **Suggestions**: The suggestions code's memory management is not great and will be replaced soon by other, more efficient methods. It is planned that the new implementation is written in C++, which avoids unnecessary overhead of the JVM and thus the memory usage should be much lower.
- **GlideTyping**: The most relevant memory leak (and really big resource hog) has already been fixed in #625, smaller memory leaks of event listeners will be fixed later.
- **LayoutManager**: The layout manager does a pretty good job at caching and preventing memory leaks, though the loading of layout json files can be improved to lower the stress on the CPU and to avoid potential garbage.
- **Re-initialization when rotating the screen**: The re-initialization of the service class when rotating the screen has some pretty nasty memory leaks and thus over the time a lot of garbage accumulates with also leads to OOM errors.
- **Settings UI**: Even though this is a small memory leak and definitely not relevant considering the garbage the above areas accumulate, the Settings UI has some small memory leaks as well which should be fixed.

---

```
java.lang.OutOfMemoryError: Failed to allocate a 24 byte allocation with 2027432 free bytes and 1979KB until OOM, target footprint 268435456, growth limit 268435456; failed due to fragmentation (largest possible contiguous allocation 24379392 bytes)
	at kotlin.reflect.jvm.internal.KParameterImpl.getType(KParameterImpl.kt:43)
	at kotlin.reflect.jvm.internal.KCallableImpl.callDefaultMethod$kotlin_reflection(KCallableImpl.kt:137)
	at kotlin.reflect.jvm.internal.KCallableImpl.callBy(KCallableImpl.kt:112)
	at com.squareup.moshi.kotlin.reflect.KotlinJsonAdapter.fromJson(KotlinJsonAdapter.kt:111)
	at com.squareup.moshi.internal.NullSafeJsonAdapter.fromJson(NullSafeJsonAdapter.java:41)
	at com.squareup.moshi.CollectionJsonAdapter.fromJson(CollectionJsonAdapter.java:81)
	at com.squareup.moshi.CollectionJsonAdapter$2.fromJson(CollectionJsonAdapter.java:55)
	at com.squareup.moshi.internal.NullSafeJsonAdapter.fromJson(NullSafeJsonAdapter.java:41)
	at com.squareup.moshi.CollectionJsonAdapter.fromJson(CollectionJsonAdapter.java:81)
```

**Comment by Flam3z** [comment]
As a long time user of your amazing app I am grateful that you put so much time into the app, listening to the community and squashing bugs like your life depends on it. Tysm

**Comment by X-yl** [comment]
I can try and reduce the glide typing memory usage a bit more. There are a few things which I think use more memory than they really should. For example, to evaluate every gesture, a new gesture is created that upsamples from the previous one. I think it could be a better idea to look into dynamically interpolating each point instead of allocating a whole new `Gesture`. Another thing could be  switching out all the `String`s for `CharArray[]`. JVM `String`s have quite an overhead, so that would probably be a good improvement. Maybe also try and replace the `FloatArray` with an `ArrayList`, but of course there's a overhead as an `ArrayList` can't store primitives, so have to profile both and see what works better.

Also regarding the suggestions, have you considered using Rust for the native part?

**Comment by Glitchy-Tozier** [comment]
@patrickgold Great idea creating this Issue.

To save yourself some work, i suggest pinning it, that might help with the flood of issues. (You could do the same with the flashing keyboard issue but this one is even more potent)

**Comment by apjyotirmay** [comment]
I'm facing an issue where the keyboard starts to lag heavily to a point where it's almost impossible to type in real time. Looks like a case of memory leak, it's happening almost over a period of 24 hours (give or take a few hours). Is this somehow related to this bug?

**Comment by Glitchy-Tozier** [comment]
@apjyotirmay Pretty sure the answer is they are connected. Whenever I rotate my phone back and fourth to make the keyboard crash, it first gets slower and only then crashes.

**Comment by Kryostar** [comment]
On version 0.3.10, turning off tbe internal clipboard and clipboard history seems makes rhe app work just fine. When clipboard is on the input has severe delay. This wasnt a problem before 0.3.10
I don't know if this is useful info but i thought if might help.

**Comment by patrickgold** [comment]
@Kryostar The clipboard manager/history wasn't really in my focus for searching memory leaks, but I will add it to the list. Could definitely also contain memory leaks which have gone undetected by now.

**Comment by felixwiemuth** [comment]
Just got a OOM crash after turning energy save mode on and off (a few times). This toggles between light and dark mode (however the keyboard first vanishes and one has to open it again manually by clicking again into a text area).

v0.3.10
Android 11

```
~~~ 1618826123214.stacktrace ~~~

java.lang.OutOfMemoryError: Failed to allocate a 524304 byte allocation with 360376 free bytes and 351KB until OOM, target footprint 268435456, growth limit 268435456
	at java.util.HashMap.resize(HashMap.java:703)
	at java.util.HashMap.putVal(HashMap.java:662)
	at java.util.HashMap.put(HashMap.java:611)
	at kotlin.collections.MapsKt__MapsKt.putAll(Maps.kt:423)
	at dev.patrickgold.florisboard.ime.text.gestures.GlideTypingManager$setWordData$1.invokeSuspend(GlideTypingManager.kt:81)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

**Comment by another-sapiens** [comment]
Suggestions seem to be the biggest issue, and gliding seems to consume a lot of resources too. 
However, I've just checked what @Kryostar said and the internal clipboard makes a huge difference in the time the keyboard takes to appear on the screen.

**Comment by another-sapiens** [comment]
Actually, scratch that. When you disable suggestions and the internal clipboard, all lag seems to be removed.

**Comment by apjyotirmay** [comment]
@another-sapiens that's right. I disabled those 2 features and the keyboard is back to being flawlessly smooth

**Comment by Glitchy-Tozier** [comment]
The question is whether it remains that way. It might just be that disabling suggestions resets some part of the memory.

**Comment by patrickgold** [comment]
> The question is whether it remains that way. It might just be that disabling suggestions resets some part of the memory.

Disabling suggestions won't actually de-allocate the memory the dictionary needs, only after a service restart the effect is truly there (as the dictionary is not loaded initially). Of course, as soon as you disable suggestions the suggestions algorithm is not triggered, improving performance vastly thus.

**Comment by LordCozyCat** [comment]
Just wanted to share my experience with this issue.

I never had suggestions turned on and over time the memory issue would grow until the keyboard would freeze and get stuck on screen.

After reading this thread I disabled the clipboard and it's much better.

**Comment by apjyotirmay** [comment]
@LordCozyCat I'm facing the same issue here. I disabled all the features and right now it's just the bare bones keyboard. However, the issue with lag, freeze, and getting stuck on screen is still there. It's taking a lot longer than when those extra features were there, but I'm still facing the issue - frequency has reduced.

The only solution for me is to reboot the device to get out of that frozen state.

**Comment by X-yl** [comment]
Hmm clipboard may be having a bigger impact than I expected. I will investigate and try to track down the memory leaks. Btw just to clarify, does it only happen if you copy and paste things (maybe lots of images or gifs?) or does it happen regardless?

@apjyotirmay you can also try to force stop the keyboard, that might help temporarily

**Comment by apjyotirmay** [comment]
I've tried that, but in my case it's more like apps go under the keyboard, and thus the force stop button is behind the keyboard when it freezes. It's not a big deal for me to reboot the device as it doesn't take a lot of my time

**Comment by patrickgold** [comment]
I've finally reached the stabilization phase of #734, which is a major rework of the layout engine and addresses core memory and performance issues (see https://github.com/florisboard/florisboard/pull/734#issuecomment-828101955 for details what is addressed and what not). If you want to try out the debug build, you can download it [here](https://github.com/florisboard/florisboard/pull/734/checks). Be aware that the touch logic is still a bit wonky and that glide typing, gestures and emojis don't work in the current debug build, which I will fix in the next two days. If you have feedback about the debug build, feel free to comment in #734! I will directly use it to improve the PR's changes before merging into master.

**Comment by Flam3z** [comment]
@patrickgold thank you so much for working on this alot, have you thought of created an element or IRC group for us users to talk to you and hopefully give you more direct feedback?

**Comment by apjyotirmay** [comment]
@Flam3z it might not be easy to keep a track on IMs. Github's features actually make all of the conversations consolidated an easy to work on

**Comment by abelian424** [comment]
Just leaving a comment of my appreciation for this app. Features have been implemented at a really fast pace, and I'll be patient waiting for upcoming fixes while I change some settings for now to mitigate the problem.

**Comment by Kreuger** [comment]
Hey Patrick. Ive been having issues with the keyboard being stuck on screen today with the beta and it only had 80mb in use. I had to switch to the stable because I kept having issues with using andotp because they would time out on me and it was making me angry. 🤣

**Comment by Huy-Ngo** [comment]
> it only had 80mb in use.

Noob question: how do you know how much memory is used on phone? Do you have some kind of monitoring app installed?

> I had to switch to the stable

Funny because I sometimes meet this issue in stable and had to switch to the debug version yesterday for the keyboard to show up. :rofl:

**Comment by Kreuger** [comment]
> > it only had 80mb in use.
> 
> Noob question: how do you know how much memory is used on phone? Do you have some kind of monitoring app installed?
> >


Theres a setting in the beta for it, in build 0.3.11-beta04. Open the 3 dot menu, hit advanced then tick the boxes "Enable developer tools" and the one below it, "show heap memory stats".

**Comment by patrickgold** [comment]
Today evening v0.3.11 will release on stable track. This release marks an important milestone in restructuring the code base + logic of FlorisBoard's text input, as well as fixing some critical parts for memory management and the by now infamous OOM errors. I've put a lot of work in this and hope that this will make FlorisBoard more usable for all users who experienced frequent OOM errors and crashes. I also want to thank @X-yl for his work on improving the glide typing code base and also fixing some other issues, @Glitchy-Tozier for the help in managing issues (especially for closing the vast amount of duplicates for this crash alone) and of course every one else who contributed either by participating in the issue discussions, creating/reviewing pull requests, translating FlorisBoard into their native language, etc. I really appreciate everyone in this awesome community :)

Next week the development cycle for v0.3.12 will begin, where I will start setting up the JNI (Java Native Interface) in order to be able to write native C++ code directly. This is mostly needed for rewriting the suggestions and implementing autocorrect & spellchecking.

**Comment by Kreuger** [comment]
Hey Patrick is there any reason why we cant enable the memory thing on the debug build?

**Comment by patrickgold** [comment]
> Hey Patrick is there any reason why we cant enable the memory thing on the debug build?

Memory thing?

**Comment by daraul** [comment]
I think I just got this, so i'm posting the trace below: 

#### Environment information
- FlorisBoard Version: 0.3.12 (43)
- Device: Xiaomi POCO F2 Pro
- Android: 11 (cn=R sdk=30)

#### Features enabled
```kt
smartbar = true
suggestions = true
suggestions_clipboard = true
suggestions_next_word = true
glide = true
clipboard_internal = true
clipboard_history = true
```

#### Attached stacktrace files
<details>
<summary>1618488417390.stacktrace</summary>

```
java.lang.OutOfMemoryError: Failed to allocate a 524304 byte allocation with 409088 free bytes and 399KB until OOM, target footprint 268435456, growth limit 268435456
	at java.util.HashMap.resize(HashMap.java:703)
	at java.util.HashMap.putVal(HashMap.java:662)
	at java.util.HashMap.put(HashMap.java:611)
	at kotlin.collections.MapsKt__MapsKt.putAll(Maps.kt:423)
	at dev.patrickgold.florisboard.ime.text.gestures.GlideTypingManager$setWordData$1.invokeSuspend(GlideTypingManager.kt:81)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)

```
</details>

<details>
<summary>1628719234982.stacktrace</summary>

```
java.util.ConcurrentModificationException
	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1441)
	at java.util.HashMap$KeyIterator.next(HashMap.java:1465)
	at dev.patrickgold.florisboard.ime.text.gestures.StatisticalGlideTypingClassifier$Pruner.<init>(StatisticalGlideTypingClassifier.kt:374)
	at dev.patrickgold.florisboard.ime.text.gestures.StatisticalGlideTypingClassifier.initializePruner(StatisticalGlideTypingClassifier.kt:138)
	at dev.patrickgold.florisboard.ime.text.gestures.StatisticalGlideTypingClassifier.setWordData(StatisticalGlideTypingClassifier.kt:121)
	at dev.patrickgold.florisboard.ime.text.gestures.GlideTypingManager$setWordData$1.invokeSuspend(GlideTypingManager.kt:80)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)

```
</details>

**Comment by egecelikci** [comment]
It happens every time I reboot, after reboot and first crash, It doesn't happen again.

#### Environment information
- FlorisBoard Version: 0.3.13 (56)
- Device: ONEPLUS A5000
- Android: 11 (cn=R sdk=30)

#### Features enabled
```kt
smartbar = true
suggestions = true
suggestions_clipboard = true
suggestions_next_word = true
glide = false
clipboard_internal = true
clipboard_history = true
```

#### Attached stacktrace files
<details>
<summary>1631157761469.stacktrace</summary>

```
java.lang.OutOfMemoryError: Failed to allocate a 56 byte allocation with 8 free bytes and 8B until OOM, target footprint 268435456, growth limit 268435456
	at android.view.inputmethod.CursorAnchorInfo$1.createFromParcel(CursorAnchorInfo.java:627)
	at android.view.inputmethod.CursorAnchorInfo$1.createFromParcel(CursorAnchorInfo.java:624)
	at com.android.internal.view.IInputMethodSession$Stub.onTransact(IInputMethodSession.java:248)
	at android.os.Binder.execTransactInternal(Binder.java:1159)
	at android.os.Binder.execTransact(Binder.java:1123)

```
</details>

<details>
<summary>1631050791331.stacktrace</summary>

```
java.lang.OutOfMemoryError: Failed to allocate a 208 byte allocation with 88 free bytes and 88B until OOM, target footprint 268435456, growth limit 268435456
	at java.util.Arrays.copyOf(Arrays.java:3257)
	at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:124)
	at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:448)
	at java.lang.StringBuilder.append(StringBuilder.java:137)
	at java.lang.Daemons$FinalizerWatchdogDaemon.finalizerTimedOut(Daemons.java:453)
	at java.lang.Daemons$FinalizerWatchdogDaemon.runInternal(Daemons.java:327)
	at java.lang.Daemons$Daemon.run(Daemons.java:139)
	at java.lang.Thread.run(Thread.java:923)

```
</details>

<details>
<summary>1631138248909.stacktrace</summary>

```
java.lang.OutOfMemoryError: OutOfMemoryError thrown while trying to throw an exception; no stack trace available

```
</details>

<details>
<summary>1631138248908.stacktrace</summary>

```
java.lang.OutOfMemoryError: Failed to allocate a 32 byte allocation with 0 free bytes and 0B until OOM, target footprint 268435456, growth limit 268435456
	at java.util.ArrayList.iterator(ArrayList.java:834)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4(temp_suggestion_code.kt:264)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4(temp_suggestion_code.kt:285)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4(temp_suggestion_code.kt:279)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4(temp_suggestion_code.kt:285)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4(temp_suggestion_code.kt:258)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4$default(temp_suggestion_code.kt:235)
	at dev.patrickgold.florisboard.ime.dictionary.FlorisLanguageModel.matchAllNgrams(temp_suggestion_code.kt:406)
	at dev.patrickgold.florisboard.ime.dictionary.Flictionary.getTokenPredictions-CIUF35U(Flictionary.kt:306)
	at dev.patrickgold.florisboard.ime.text.TextInputManager$onWordHistoryChanged$1.invokeSuspend(TextInputManager.kt:1010)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)

```
</details>

<details>
<summary>1631572230265.stacktrace</summary>

```
java.util.concurrent.TimeoutException: com.android.internal.os.BinderInternal$GcWatcher.finalize() timed out after 10 seconds
	at java.lang.ref.FinalizerReference.add(FinalizerReference.java:61)
	at com.android.internal.os.BinderInternal$GcWatcher.finalize(BinderInternal.java:63)
	at java.lang.Daemons$FinalizerDaemon.doFinalize(Daemons.java:291)
	at java.lang.Daemons$FinalizerDaemon.runInternal(Daemons.java:278)
	at java.lang.Daemons$Daemon.run(Daemons.java:139)
	at java.lang.Thread.run(Thread.java:923)

```
</details>

**Comment by Glitchy-Tozier** [comment]
@patrickgold could we decide on and pin a main OOM-Issue to help with the many reports?

**Comment by patrickgold** [comment]
I could pin this issue for the OOM issues, as this is already an issue with lots of discussion around this topic.

**Comment by Glitchy-Tozier** [comment]
I think the pinned OOM issue should also have a fitting first message. Something that explains: It's only temporary ... wait till 0.3.15... disable suggestions... etc.

**Comment by TheLoving** [comment]
Description: 
Sometimes when using it freezes and stays on the screen (keyboard) for 5 or more seconds.

Steps to reproduce:
I cannot say anything as it is very random and frequently (mostly today) occuring.

#### Environment information
- FlorisBoard Version: 0.3.13 (56)
- Device: OPPO CPH1859
- Android: 9 (cn=P sdk=28)

#### Features enabled
```kt
smartbar = true
suggestions = true
suggestions_clipboard = true
suggestions_next_word = true
glide = false
clipboard_internal = true
clipboard_history = true
```

#### Attached stacktrace files
<details>
<summary>1636720529297.stacktrace</summary>

```
java.lang.OutOfMemoryError: Failed to allocate a 80 byte allocation with 8 free bytes and 8B until OOM, max allowed footprint 402653184, growth limit 402653184
	at java.util.Arrays.copyOf(Arrays.java:3260)
	at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:125)
	at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:451)
	at java.lang.StringBuilder.append(StringBuilder.java:137)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.setIndexInArray(CoroutineScheduler.kt:589)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.tryTerminateWorker(CoroutineScheduler.kt:842)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.park(CoroutineScheduler.kt:800)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.tryPark(CoroutineScheduler.kt:740)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:711)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)

```
</details>

<details>
<summary>1636720529368.stacktrace</summary>

```
java.lang.OutOfMemoryError: Failed to allocate a 32 byte allocation with 8 free bytes and 8B until OOM, max allowed footprint 402653184, growth limit 402653184
	at java.util.ArrayList.iterator(ArrayList.java:834)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4(temp_suggestion_code.kt:264)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4(temp_suggestion_code.kt:264)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4(temp_suggestion_code.kt:291)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4(temp_suggestion_code.kt:285)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4(temp_suggestion_code.kt:258)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4$default(temp_suggestion_code.kt:235)
	at dev.patrickgold.florisboard.ime.dictionary.FlorisLanguageModel.matchAllNgrams(temp_suggestion_code.kt:406)
	at dev.patrickgold.florisboard.ime.dictionary.Flictionary.getTokenPredictions-CIUF35U(Flictionary.kt:306)
	at dev.patrickgold.florisboard.ime.text.TextInputManager$onWordHistoryChanged$1.invokeSuspend(TextInputManager.kt:1010)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)

```
</details>

<details>
<summary>1636720537830.stacktrace</summary>

```
java.lang.OutOfMemoryError: Failed to allocate a 32 byte allocation with 8 free bytes and 8B until OOM, max allowed footprint 402653184, growth limit 402653184
	at java.util.ArrayList.iterator(ArrayList.java:834)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4(temp_suggestion_code.kt:264)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4(temp_suggestion_code.kt:285)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4(temp_suggestion_code.kt:291)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4(temp_suggestion_code.kt:279)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4(temp_suggestion_code.kt:258)
	at dev.patrickgold.florisboard.ime.dictionary.NgramNode.listSimilarWords-qoaOXc4$default(temp_suggestion_code.kt:235)
	at dev.patrickgold.florisboard.ime.dictionary.FlorisLanguageModel.matchAllNgrams(temp_suggestion_code.kt:406)
	at dev.patrickgold.florisboard.ime.dictionary.Flictionary.getTokenPredictions-CIUF35U(Flictionary.kt:306)
	at dev.patrickgold.florisboard.ime.text.TextInputManager$onWordHistoryChanged$1.invokeSuspend(TextInputManager.kt:1010)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)

```
</details>

<details>
<summary>1636720542123.stacktrace</summary>

```
java.lang.OutOfMemoryError: Failed to allocate a 24 byte allocation with 8 free bytes and 8B until OOM, max allowed footprint 402653184, growth limit 402653184
	at com.mediatek.libcore.MTKStringFactory.newStringFromUTF8Bytes(Native Method)
	at java.lang.StringFactory.newStringFromBytes(StringFactory.java:88)
	at dev.patrickgold.florisboard.ime.dictionary.Flictionary$Companion.load-B7oiUAE(Flictionary.kt:134)
	at dev.patrickgold.florisboard.ime.dictionary.DictionaryManager.loadDictionary-pqiQ7go(DictionaryManager.kt:94)
	at dev.patrickgold.florisboard.ime.text.TextInputManager$onWordHistoryChanged$1.invokeSuspend(TextInputManager.kt:1009)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)

```
</details>

<details>
<summary>1636720549965.stacktrace</summary>

```
java.lang.OutOfMemoryError: OutOfMemoryError thrown while trying to throw OutOfMemoryError; no stack trace available

```
</details>

<details>
<summary>1636720552605.stacktrace</summary>

```
java.lang.OutOfMemoryError: Failed to allocate a 176 byte allocation with 8 free bytes and 8B until OOM, max allowed footprint 402653184, growth limit 402653184
	at java.lang.StringBuilder.toString(StringBuilder.java:410)
	at android.os.Looper.loop(Looper.java:241)
	at android.app.ActivityThread.main(ActivityThread.java:7178)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:503)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:942)

```
</details>

**Comment by lm41** [comment]
@patrickgold Is the first bullet point still relevant?

**Comment by patrickgold** [comment]
> patrickgold Is the first bullet point still relevant? 

Nope isn't, that problematic section is not part of the code base anymore

**Comment by tsiflimagas** [comment]
> Any updates to this?

Well, I think none of these memory leaks is there anymore. Either they've been fixed, or the part where they appeared has been removed or disabled.

@patrickgold it seems like this issue can be closed, or is there still a need for it?

**Comment by patrickgold** [comment]
In the settings UI, what? That's not really the area this crash usually occurred
